ethereum/contracts/BeefyClient.sol

```diff
@@ -0,0 +1,474 @@
+
+        // drop admin privileges
+        renounceRole(DEFAULT_ADMIN_ROLE, msg.sender);
+    }
+
+    /* Public Functions */
+
+    /**
+     * @notice Executed by the incoming channel in order to verify commitment
+     * @param leafHash contains the merkle leaf to be verified
+     * @param proof contains simplified mmr proof
+     */
+    function verifyMMRLeafProof(bytes32 leafHash, MMRProof calldata proof)
+        external
+        view
+        returns (bool)
+    {
+        return MMRProofVerification.verifyLeafProof(latestMMRRoot, leafHash, proof);
+    }
+
+    /**
+     * @notice Executed by the prover in order to begin the process of block
+     * acceptance by the light client
+     * @param commitmentHash contains the commitmentHash signed by the validator(s)
+     * @param validatorClaimsBitfield a bitfield containing a membership status of each
+     * validator who has claimed to have signed the commitmentHash
+     * @param validatorSignature the signature of one validator
+     * @param validatorPosition the position of the validator, index starting at 0
+     * @param validatorPublicKey the public key of the validator
+     * @param validatorPublicKeyMerkleProof proof required for validation of the public key in the validator merkle tree
+     */
+    function newSignatureCommitment(
+        bytes32 commitmentHash,
+        uint64 validatorSetID,
+        uint256[] memory validatorClaimsBitfield,
+        bytes memory validatorSignature,
+        uint256 validatorPosition,
+        address validatorPublicKey,
+        bytes32[] calldata validatorPublicKeyMerkleProof
+    ) public payable {
+        ValidatorSet memory vset = getValidatorSet(validatorSetID);
+
+        // Check if validatorPublicKeyMerkleProof is valid based on validatorSetRoot
+        require(
+            isValidatorInSet(
+                vset,
+                validatorPublicKey,
+                validatorPosition,
+                validatorPublicKeyMerkleProof
+            ),
+            "Sender must be in validator set at correct position"
+        );
+
+        // Check if validatorSignature is correct, ie. check if it matches
+        // the signature of senderPublicKey on the commitmentHash
+        require(
+            ECDSA.recover(commitmentHash, validatorSignature) == validatorPublicKey,
+            "Invalid Signature"
+        );
+
+        // Check that the bitfield actually contains enough claims to be successful, ie, >= 2/3
+        require(
+            validatorClaimsBitfield.countSetBits() >= requiredNumberOfSignatures(vset),
+            "Not enough claims in bitfield"
+        );
+
+        // Accept and save the commitment
+        validationData[nextID] = ValidationData(
+            msg.sender,
+            commitmentHash,
+            vset.id,
+            validatorClaimsBitfield,
+            block.number
+        );
+
+        emit CommitmentVerified(nextID, Phase.Initial, commitmentHash, msg.sender);
```

Comment:
why choose to use the same event for both initial and final verification? seems a bit risky and may also not be ideal for consumers who only care about final commitments?

Reply:
Totally reworked in the latest code. Now a single `NewRequest` is emitted after initial verification, and nothing else is seems necessary.